### PR TITLE
BlazorTemplate test improvements

### LINKED ISSUE #26679: Add LoggedTest to Components E2E and Blazor template tests
The LoggedTest infrastructure is now available to Components E2E and Blazor template tests, however it isn't being used yet as that probably requires some knowledge around those tests to use it correctly.

Here is the change to fix the infrastructure and add the logging to the other template tests.
https://github.com/dotnet/aspnetcore/pull/26507

### PR DESCRIPTION
**Pull request by HaoK**
Try to split up the template tests, move create/build/publish away from the run

### CODE REVIEW

**Comment by HaoK** [comment]
Failures: so far 1 out 5

```
[xUnit.net 00:34:07.13] BlazorTemplates.Tests: [Long Running Test] 'Templates.Test.BlazorServerTemplateTest.BlazorServerTemplateWorks_NoAuth(browserKind: Chromium)', Elapsed: 00:01:00
  Passed Templates.Test.BlazorServerTemplateTest.BlazorServerTemplateWorks_NoAuth(browserKind: Chromium) [1 m 9 s]
  Passed AlwaysTestTests.SuccessfulTests.GuaranteedUnquarantinedTest [1 ms]
['BlazorTemplates.Tests--net6.0' WORK ITEM LOG TRUNCATED: Command timed out, and was killed]
```

**Comment by HaoK** [comment]
Summary of current improvements:
- Switch to per test instance of Playwright to prevent cross test contamination (previously playwright browser instances were shared via a shared test fixture)
- Split out create/build/publish into their own tests, and the running app/playwright part as a separate step
- Move logging of tests into its own file
- Tweak the playwright debug output to only show up in the console log for tests that fail (instead of always like before)

File: src/ProjectTemplates/BlazorTemplates.Tests/BlazorServerTemplateTest.cs [hunk 1]
```diff
@@ -17,62 +16,49 @@
...
+
+        [Theory, TestPriority(BUILDCREATEPUBLISH_PRIORITY)]
+        [MemberData(nameof(BlazorServerTemplateWorks_IndividualAuthData))]
+        public Task BlazorServerTemplate_CreateBuildPublish_IndividualAuthUseLocalDb(BrowserKind browserKind, bool useLocalDB)
+            => CreateBuildPublishAsync("blazorserverindividual" + browserKind + (useLocalDB ? "uld" : ""));
 
         [Theory]
         [InlineData(BrowserKind.Chromium)]
-        [QuarantinedTest("https://github.com/dotnet/aspnetcore/issues/30761")]
+        //[QuarantinedTest("https://github.com/dotnet/aspnetcore/issues/30761")]
```
Comment (HaoK) [inline src/ProjectTemplates/BlazorTemplates.Tests/BlazorServerTemplateTest.cs:40]: Will requarantine all tests before merge

File: src/ProjectTemplates/BlazorTemplates.Tests/BlazorWasmTemplateTest.cs [hunk 2]
```diff
@@ -131,6 +125,8 @@ public async Task BlazorWasmHostedTemplate_Works(BrowserKind browserKind)
             {
                 EnsureBrowserAvailable(browserKind);
             }
+            
+            Assert.True(1 == 0);
```
Comment (javiercn) [inline src/ProjectTemplates/BlazorTemplates.Tests/BlazorWasmTemplateTest.cs:129]: ??

Comment (HaoK) [reply src/ProjectTemplates/BlazorTemplates.Tests/BlazorWasmTemplateTest.cs:129]: Its really hard to see the logs without a failing test in the run UI, so I always break a test, the last thing I'm trying to chase down is where the webapp output went, I got the playwright logging to show up in the log file, but the debug output from the server is gone now

File: src/Testing/src/LoggedTest/LoggedTestBase.cs [hunk 3]
```diff
@@ -95,6 +95,12 @@ public virtual void Initialize(TestContext context, MethodInfo methodInfo, objec
                 _initializationException = ExceptionDispatchInfo.Capture(e);
             }
         }
+        
+        public virtual Task InitializeAsync(TestContext context, MethodInfo methodInfo, object[] testMethodArguments, ITestOutputHelper testOutputHelper)
+        {
+            Initialize(context, methodInfo, testMethodArguments, testOutputHelper);
+            return Task.CompletedTask;
+        }
```
Comment (javiercn) [inline src/Testing/src/LoggedTest/LoggedTestBase.cs:104]:
```suggestion
        public virtual Task InitializeAsync(TestContext context, MethodInfo methodInfo, object[] testMethodArguments, ITestOutputHelper testOutputHelper)
        {
            Initialize(context, methodInfo, testMethodArguments, testOutputHelper);
            return InitializeCoreAsync(context);
        }
        
        protected virtual Task InitializeCoreAsync() => Task.CompletedTask;
```
We do this in other places, the advantage is that you don't have to remember to call base.

**Comment by HaoK** [comment]
@javiercn server logs are finally working now, so other than all the tests that need to be requarantined I think this is ready for final review

File: src/ProjectTemplates/BlazorTemplates.Tests/TestPriorityAttribute.cs [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+
+namespace Templates.Test
+{
+    [AttributeUsage(AttributeTargets.Method, AllowMultiple = false)]
+    public class TestPriorityAttribute : Attribute
+    {
+        public int Priority { get; private set; }
+        public TestPriorityAttribute(int priority) => Priority = priority;
+    }
+
```
Comment (javiercn) [inline src/ProjectTemplates/BlazorTemplates.Tests/TestPriorityAttribute.cs:11]: nit: extra line

File: src/ProjectTemplates/BlazorTemplates.Tests/BlazorTemplateTest.cs [hunk 5]
```diff
@@ -1,49 +1,146 @@
...
+            }
         }
+        
+        public abstract string ProjectType { get; }
+        private static readonly bool _isCIEnvironment =
+            !string.IsNullOrWhiteSpace(Environment.GetEnvironmentVariable("ContinuousIntegrationBuild"));
+
+        protected async override Task InitializeCoreAsync(TestContext context)
+        {
+            BrowserManager = await BrowserManager.CreateAsync(CreateConfiguration(), LoggerFactory);
```
Comment (javiercn) [inline src/ProjectTemplates/BlazorTemplates.Tests/BlazorTemplateTest.cs:53]: As per discussion, you will undo this after we get this in? (revert to it being a single fixture)

Comment (HaoK) [reply src/ProjectTemplates/BlazorTemplates.Tests/BlazorTemplateTest.cs:53]: Yep I'll file an issue, its got a complication in that the new nice file logging per test only works well as a side effect of this split too, so its going to be not completely trivial

Comment (javiercn) [reply src/ProjectTemplates/BlazorTemplates.Tests/BlazorTemplateTest.cs:53]: I'm fine with that, as long as we push all the playwright logs into its own debug file, that's good enough for me.

Comment (HaoK) [reply src/ProjectTemplates/BlazorTemplates.Tests/BlazorTemplateTest.cs:53]: https://github.com/dotnet/aspnetcore/issues/31111

**Comment by HaoK** [comment]
Any outstanding issues that we need to address in this PR @javiercn ?  I'd like to get this in so we can work on more improvements + refactor things to share with the new p2w version of tests

**Comment by javiercn** [comment]
@HaoK let me take a final look, I've only perused through it

**Review by javiercn — APPROVED** [review]
Looks great!

**Comment by BrennanConroy** [comment]
I think this change made the tests really flaky
image

**Comment by HaoK** [comment]
@javiercn I'm going to split this PR into two separate changes I guess, the first one that should be safe is the logging changes.  I'm going to leave the tests alone for now, since it looks like the splitting of tests into create/build/hosting with the orderer might be the issue, the restore might not have worked.

**Comment by HaoK** [comment]
Looks like the underlying issues with the create/build split is that the restore step seems to be more flaky after the change

**Comment by javiercn** [comment]
@HaoK do we have an idea why that would be the case for restore?

**Comment by dougbu** [comment]
> Looks like the underlying issues with the create/build split is that the restore step seems to be more flaky after the change

Is this related to our persistent-but-not-consistent DNS failures @HaoK❔

**Comment by javiercn** [comment]
@dougbu could be?

I can't see a good reason where the changes hao made could make restore unstable

**Comment by HaoK** [comment]
In this PR there's not really any real changes in the build/create code other than moving them to a shared helper, they are called as part of the test same as before, so it doesn't look like the splitting of the tests into two parts was the source of the instability.  These tests have pretty much all been quarantined for a while so its possible things have been broken by something else earlier

**Comment by HaoK** [comment]
I'm not sure if there's any difference between the blazor templates and the project templates, since i'd assume those are not running into these issues

**Comment by HaoK** [comment]
Does anyone know why we need to do --no-restore when we run these tests since they are inside of dotnet locks?  I'm going to try removing the --no-restore just to see if this helps at all

**Pull request merged**